### ISSUE TITLE
'Max call stack exceeded' error

### ISSUE THREAD

**Issue by jpellizzari** [weaveworks/scope#2293]
Steps to reproduce: 

1. ~~Load https://frontend.dev.weave.works~~ (see url below)
2. Check the JS console

Expected:
Everything works

Actual: 
Max call stack error and super slowness

**Comment by jpellizzari** [comment]
Initial hunch is that this is causing problems: 
https://github.com/weaveworks/scope/commit/4f24d9bba8743bf389798f16fe56905d67ed6d93#diff-5897bc396ca616224f48758015a6ece1R215

**Comment by jpellizzari** [comment]
Seems to be localized to this URL state:

https://frontend.dev.weave.works/app/proud-wind-05#!/state/{"controlPipe":{"id":"pipe-823277968531583873","nodeId":"02c4fa6bc09fef2e97e0ba68f2a56f5159e861399c951f7c953badc1b8152a63;<container>","raw":true,"resizeTtyControl":{"id":"docker_resize_exec_tty","probeId":"3616cce1df0a057e","nodeId":"02c4fa6bc09fef2e97e0ba68f2a56f5159e861399c951f7c953badc1b8152a63;<container>"},"control":{"probeId":"3616cce1df0a057e","nodeId":"02c4fa6bc09fef2e97e0ba68f2a56f5159e861399c951f7c953badc1b8152a63;<container>","id":"docker_exec_container","human":"Exec shell","icon":"fa-terminal","rank":2}},"topologyViewMode":"topo","nodeDetails":[{"id":"02c4fa6bc09fef2e97e0ba68f2a56f5159e861399c951f7c953badc1b8152a63;<container>","label":"tom","topologyId":"containers"}],"pinnedMetricType":"CPU","pinnedSearches":[],"searchQuery":"","selectedNodeId":"02c4fa6bc09fef2e97e0ba68f2a56f5159e861399c951f7c953badc1b8152a63;<container>","gridSortedBy":"host_scope_version","gridSortedDesc":null,"topologyId":"containers","topologyOptions":{"processes":{"unconnected":"hide"},"containers":{"system":"all","stopped":"running","pseudo":"hide"},"pods":{"namespace":"scope","pseudo":"hide"}},"contrastMode":true}

**Comment by jpellizzari** [comment]
The root cause appears to be the redux `dispatch` going crazy. When you log `action.type` to console:
![screen shot 2017-03-02 at 10 51 09 am](https://cloud.githubusercontent.com/assets/2802257/23521990/3c9c43a2-ff36-11e6-8c04-47db9c7f5265.png)

**Comment by rade** [comment]
*interesting*. Yes, I did have contrast mode on.

**Comment by jpellizzari** [comment]
The root cause of this bug is the `shouldReplaceState` function using the presence of controlPipe in the URL state to determine whether or not to do a browser history 'replace'. Doing a replace triggers a re-route, which triggers the contrast mode updateRoute, which triggers a re-route, and so on.

```javascript
function shouldReplaceState(prevState, nextState) {
  // Opening a new terminal while an existing one is open.
  const terminalToTerminal = (prevState.controlPipe && nextState.controlPipe);
  // Closing a terminal.
  const closingTheTerminal = (prevState.controlPipe && !nextState.controlPipe);
  return terminalToTerminal || closingTheTerminal;
}
```

So when a user has contrast mode turned on and has a control pipe open, reloading the page will cause this issue.

**Comment by jpellizzari** [comment]
@foot Do you recall why this needs to be a `replace` instead of a `show`?:
https://github.com/weaveworks/scope/blob/3b2af2ae6c848f4772515525afcdab11d706991a/client/app/scripts/utils/router-utils.js#L25

I tried moving the check for `contrastMode` out of the `route` action, but there is no other place that has access to new URL states. 

Another option would be removing contrastMode from the URL state. The page would revert to normal mode on refresh. I like saving it, however.

**Comment by rade** [comment]
Fixed by #2294 but introduces #2295.

**Comment by davkal** [comment]
@jpellizzari 
> Do you recall why this needs to be a replace instead of a show?:

From the blamed commit:

- You can reload a terminal but it doesn't exist in history stack.
- History hacking wasn't working, this is a little simpler.
